pkg/lock/lock_debug.go

```diff
@@ -28,8 +43,89 @@ func init() {
 
 type internalRWMutex struct {
 	deadlock.RWMutex
+	time.Time
+}
+
+func (i *internalRWMutex) Lock() {
+	i.RWMutex.Lock()
+	i.Time = time.Now()
+}
+
+func (i *internalRWMutex) Unlock() {
+	if sec := time.Since(i.Time).Seconds(); sec >= selfishThresholdSec {
+		printStackTo(sec, debug.Stack(), os.Stderr)
+	}
+	i.RWMutex.Unlock()
+}
+
+func (i *internalRWMutex) UnlockIgnoreTime() {
+	i.RWMutex.Unlock()
+}
+
+func (i *internalRWMutex) RLock() {
+	i.RWMutex.Lock()
+	i.Time = time.Now()
+}
+
+func (i *internalRWMutex) RUnlock() {
+	if sec := time.Since(i.Time).Seconds(); sec >= selfishThresholdSec {
+		printStackTo(sec, debug.Stack(), os.Stderr)
+	}
+	i.RWMutex.Unlock()
+}
+
+func (i *internalRWMutex) RUnlockIgnoreTime() {
+	i.RWMutex.Unlock()
 }
 
 type internalMutex struct {
 	deadlock.Mutex
+	time.Time
+}
+
+func (i *internalMutex) Lock() {
+	i.Mutex.Lock()
+	i.Time = time.Now()
+}
+
+func (i *internalMutex) Unlock() {
+	if sec := time.Since(i.Time).Seconds(); sec >= selfishThresholdSec {
+		printStackTo(sec, debug.Stack(), os.Stderr)
+	}
+	i.Mutex.Unlock()
+}
+
+func (i *internalMutex) UnlockIgnoreTime() {
+	i.Mutex.Unlock()
+}
+
+func printStackTo(sec float64, stack []byte, writer io.Writer) {
+	goRoutineNumber := []byte("0")
+	newLines := 0
+
+	if bytes.Equal([]byte("goroutine"), stack[:len("goroutine")]) {
+		newLines = bytes.Count(stack, []byte{'\n'})
+		goroutineLine := bytes.IndexRune(stack, '[')
+		goRoutineNumber = stack[:goroutineLine]
+	}
+
+	log.WithFields(logrus.Fields{
+		"seconds":   sec,
+		"goroutine": string(goRoutineNumber[len("goroutine") : len(goRoutineNumber)-1]),
+	}).Debugf("Took more than %.2f seconds", selfishThresholdSec)
```

Comment:
Two things here: 

- Could you add this message to `helpers.checkLogsMessages`
- Should the messages be a warning or info?
- The log message for me should be `Goroutine lock took more than` so is easy to read, I know that has a subsys, but I think that it's easy to read.

Reply:
>Two things here:
>
You mentioned 3 things. Out of bounds exception.

>* Could you add this message to helpers.checkLogsMessages
Done
>* Should the messages be a warning or info?
No, they will print the trace call as well. It will be to verbose IMO
>* The log message for me should be Goroutine lock took more than so is easy to read, I know that has a subsys, but I think that it's easy to read.
Done